plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/expression/SparkExpressionBuilder.java

```diff
@@ -0,0 +1,231 @@
+                        unicodeStringBuilder.append(ch);
+                    }
+                    break;
+                case ESCAPED:
+                    if (ch == SPARK_ESCAPE_CHARACTER) {
+                        unicodeStringBuilder.append(SPARK_ESCAPE_CHARACTER);
+                        state = UnicodeDecodeState.EMPTY;
+                    }
+                    else if (ch == 'u') {
+                        state = UnicodeDecodeState.UNICODE_SEQUENCE;
+                        charactersNeeded = 4;
+                    }
+                    else if (ch == 'U') {
+                        state = UnicodeDecodeState.UNICODE_SEQUENCE;
+                        charactersNeeded = 8;
+                    }
+                    else if (isHexDigit(ch)) {
+                        state = UnicodeDecodeState.UNICODE_SEQUENCE;
+                        escapedCharacterBuilder.append(ch);
+                    }
+                    else {
+                        throw new ParsingException("Invalid hexadecimal digit: " + ch);
+                    }
+                    break;
+                case UNICODE_SEQUENCE:
+                    checkState(isHexDigit(ch), "Incomplete escape sequence: " + escapedCharacterBuilder, context);
+                    escapedCharacterBuilder.append(ch);
+                    if (charactersNeeded == escapedCharacterBuilder.length()) {
+                        String currentEscapedCode = escapedCharacterBuilder.toString();
+                        escapedCharacterBuilder.setLength(0);
+                        int codePoint = Integer.parseInt(currentEscapedCode, 16);
+                        checkState(Character.isValidCodePoint(codePoint), "Invalid escaped character: " + currentEscapedCode, context);
+                        if (Character.isSupplementaryCodePoint(codePoint)) {
+                            unicodeStringBuilder.appendCodePoint(codePoint);
+                        }
+                        else {
+                            char currentCodePoint = (char) codePoint;
+                            checkState(!Character.isSurrogate(currentCodePoint), format("Invalid escaped character: %s. Escaped character is a surrogate. Use '\\+123456' instead.", currentEscapedCode), context);
+                            unicodeStringBuilder.append(currentCodePoint);
+                        }
+                        state = UnicodeDecodeState.EMPTY;
+                        charactersNeeded = -1;
+                    }
+                    else {
+                        checkState(charactersNeeded > escapedCharacterBuilder.length(), "Unexpected escape sequence length: " + escapedCharacterBuilder.length(), context);
+                    }
+                    break;
+                default:
+                    throw new UnsupportedOperationException();
+            }
+        }
+
+        checkState(state == UnicodeDecodeState.EMPTY, "Incomplete escape sequence: " + escapedCharacterBuilder, context);
```

Comment:
That probably should be a ParseException.

Also, maybe worth providing the offending literal in full too?

```java
format("Incomplete escape sequence '%s' at the end of %s literal", escapedCharacterBuilder, ontext.getText())
```